dln/loss.py

```diff
@@ -1,36 +1,113 @@
+        if not hasattr(cls, '_available_losses'):
+            cls._available_losses = {}
+        else:
+            cls._available_losses[name] = cls
 
 
-class ZeroOneLoss(LLoss):
-    def __init__(self, postproc=None):
+class LLoss(ABC, metaclass=LLossMeta):
+
+    @classmethod
+    def available_losses(cls):
+        return list(cls._available_losses.keys())
+
+    @classmethod
+    def instantiate(cls, loss_type: str, postproc: Optional[Callable] = None) -> "LLoss":
+        try:
+            return cls._available_losses[loss_type](postproc)
+        except KeyError:
+            raise ValueError(f'Unknown loss type: {loss_type}')
+
+    def __init__(self, postproc: Optional[Callable] = None):
         """
         Args:
             postproc: a function that takes and returns a string to be apply before calculating the loss
-        Returns:
-            ZeroOneLoss as an float32 np.array
         """
         self._postproc = postproc
 
-    def __call__(self, input, target):
+    @property
+    def postproc(self):
+        """
+        Returns the post-processing function for the loss function.
+        If the post-processing function has not been set, returns the identity function
+        """
+        if self._postproc is None:
+            return lambda x: x
+        return self._postproc
+
+    @abstractmethod
+    def loss(self, inputs: Iterable[str], targets: Iterable[str]) -> np.array:
         """
+        Computes the loss between the input and target
         Args:
-            input: a list of strings
-            target: a list of strings
+            input: The predicted outputs
+            target: The true outputs
+        Returns:
+            The computed loss
         """
+        pass
+
+    def __call__(
+        self,
+        inputs: Union[str, Iterable[str]],
+        targets: Union[str, Iterable[str]],
+    ) -> np.array:
+        """
+        Calls the loss function. If inputs or targets are not iterables, they are converted to lists
+        Args:
+            inputs: The predicted outputs
+            targets: The true outputs
+        Returns:
+            The computed loss as an np.array
+        """
+        if isinstance(inputs, str) or not isinstance(inputs, Iterable):
+            inputs = [inputs]
+        if isinstance(targets, str) or not isinstance(targets, Iterable):
+            targets = [targets]
         if self._postproc:
-            input = [self.postproc(i) for i in input]
-            target = [self.postproc(t) for t in target]
-        losses = (np.array(input) != np.array(target)).astype("float32")
+            inputs = [self.postproc(i) for i in inputs]
+            targets = [self.postproc(t) for t in targets]
+        losses = self.loss(inputs, targets)
         return losses
 
-    @property
-    def postproc(self):
-        if self._postproc is None:
-            return lambda x: x
-        return self._postproc
\ No newline at end of file
+
+class ZeroOneLoss(LLoss):
```

Comment:
I believe a better name for this one would be ExactMatch.

Reply:
Nice, thanks for the suggestion! I am changing this name in the next commit.